rfcs/detecting_sensitive_textual_content/20230309-project_proposal_detecting_sensitive_textual_content.md

```diff
@@ -0,0 +1,438 @@
+## Project summary
+
+Detect results with sensitive textual content by checking their textual content
+against a list of sensitive terms. Once results with sensitive terms in their
+textual content are determined, blur those results on the frontend.
+
+> **Note** We acknowledge that we are making an inference about results based on
+> textual context. That inference applies specifically to the textual content
+> _only_, but is also likely to apply to content itself (whether image or
+> audio). We also acknowledge the various pit-falls that can come with this
+> approach as well as its insufficency. However, it's the best approach we can
+> implement now, in a timely manner, that will substantively improve the
+> accessibility of Openverse.
+
+## Goals
+
+This advances Openverse's goal of increasing the safety of the platform.
+
+## Requirements
+
+> **A note on tone**: This section should be mostly readable and understandable
+> by casual, non-technical readers. However, certain aspects of it unavoidably
+> delve into technical details in order to describe how features fit together or
+> considerations that will need to be kept in mind during implementation
+> planning. These need to be documented _somewhere_, and displacing them from
+> the requirements felt strange. If you are not interested in these particular
+> technical details, please skip the two sections titled "Implementation
+> details" and the section on "How to blur results". These sections in
+> particular consider technical issues of the project that may confound readers
+> unfamiliar with software development whether applied to Openverse or more
+> generally.
+
+### API
+
+#### Filtering
+
+The Django API is able to filter out results that include sensitive terms in
+their textual content. It does so without degrading search performance.
+
+Filtering of results with sensitive terms in their textual content occurs when
+`include_sensitive_results=True`. This parameter will supplant the `mature=True`
+as a more comprehensively and descriptively named parameter. `mature=True`
+should still work, but it should just be an alias for
+`include_sensitive_results` and should be marked as deprecated in or altogether
+excluded from the API documentation.
+
+##### Implementation details
+
+This will likely be implemented as a secondary index of results that do not
+include the sensitive terms, as explored in
+https://github.com/WordPress/openverse-api/pull/1108.
+
+#### Designation of results with sensitive terms
+
+Results that include sensitive terms in their textual content are so denoted in
+the API results in a new result field named `sensitivity`. Sensitivity should be
+an array that should be empty for results where `mature = false` and the textual
+content does not contain any sensitive terms. For results marked as `mature`,
+the array should include a `"mature"` string. For results with sensitive terms
+in their textual content, the array should include a `"sensitive_text"` string.
```

Comment:
This is speaking of the Response serializers, which sounds great to me. However, do you know how the manually reported content will fit into this process? For example, will sensitive terms be included as tags? Do we need to add more columns? Will the `api_mature<media>` tables may be deprecated?

Reply:
I hadn't considered the mature reports and had to look at the code to learn how they worked again, specifically, how the information gets incorporated into search. Forgive me if you already know this, but it was new to me and to understand my answer this is necessary context, so for my sake and for others who might not be familiar I am including it.

When the mature report is created, the [`.save` method updates the document in ES](https://github.com/wordpress/openverse/blob/7c92ba5bdff84b2771dc25a946bb371600906e0e/api/catalog/api/models/media.py#L362) to set `mature=True` and forces a refresh of the index.

There are a couple details here:

1. This will automatically work with the proposed solution, the reports just get folded into the rest of the documents already labelled as "mature". These documents would have the "mature" string in the sensitivity array.
2. I believe this is slightly broken in that this information is not preserved between data refreshes. As far as I can tell, there is no mechanism to re-apply the setting of `mature=true` for results that have confirmed reports during data refresh. Once a new index is created, the mature setting on those documents will go back to `false`. I tried digging for something that would prevent this, but I can't find it. If you take a look at the code and confirm my suspicion, I'll open an issue for us to investigate how to fix this. I suspect we would need to use a similar backport strategy as the others in discussion to get the confirmed reports from API to catalogue. We could also use Elasticsearch ingestion pipelines to flip the mature flag on documents as they're being ingested based on the mature report tables in the API database. Yet another alternative is to iterate through the mature report table after indexing and set the flag on the documents in a batch update. All are probably complex solutions to implement with their own pitfalls that would need to be considered.

So that being said, if the mature content reporting process didn't have that data refresh issue, I think it would work perfectly without modification with this proposal. If we fix that issue, then it will work with the proposal, but there's nothing this proposal needs to do to handle that bug. It is a separate bug and doesn't impede the sensitive terms detection discussed here. We would keep the mature tables around, provided we don't decide a different way of handling them is necessary in the course of fixing the bug described.

> will sensitive terms be included as tags? Do we need to add more columns?

Do you mean tags on the Elasticsearch documents? I don't believe it's necessary and would make search more complicated as we'd have to somehow prevent those sensitive terms from affecting the overall score of the document. Each tag would be adding an instance of the term on the document's overall textual content.

As for adding new columns, I don't believe it's necessary. Right now we're not classifying the particular terms into categories, so the only token that would exist in the sensitivity array is `sensitive_text`.

I guess it is worth clarifying explicitly that documents labelled "mature" are separate entirely (at least conceptually) from those we find have potentially sensitive textual content.

Reply:
Great call about considering the data refresh pieces of this! I looked at the code, and it appears that when building the index we **do already consider** deleted and mature records:

https://github.com/WordPress/openverse/blob/0f90f84d4bf856297356a918e4cc4fe42d07ffc2/ingestion_server/ingestion_server/queries.py#L28-L38

That function is called in the update & reindex steps, and I believe the `mature` value returned there is applied to the index. So if the record exists in `api_mature{media}`, it will have `mature=True` in the resultant index:

https://github.com/WordPress/openverse/blob/0f90f84d4bf856297356a918e4cc4fe42d07ffc2/ingestion_server/ingestion_server/indexer_worker.py#L58-L60

I was unsure about this myself because my initial assumption mirrored your own Sara, I'm glad that @dhruvkb ensured that we handled this appropriately during the refresh 😄